main.go

```diff
@@ -396,6 +511,187 @@ func runInPod(name string, cmdToRun string, env []string, timeout int) ([]string
+	} else if step.Selection != nil {
+		nodes = selectNodesFromSelection(step, config, subsetPartition)
+	} else {
+		fatal("Unexpected selection input")
+	}
+	return nodes
+}
+
+func selectNodesFromSelection(step Step, config Config, subsetPartition map[int][]int) []int{
+	if (step.Selection.Range == nil && step.Selection.Percent == nil && step.Selection.Subset == nil) {
+		fatal("No selection method on test step")
+	} else if (step.Selection.Range != nil) {
+		return selectNodesRange(step, config)
+	} else if (step.Selection.Percent != nil) {
+		return selectNodesPercent(step, config)
+	}
+	/* Subset selection */ 
+	return selectNodesSubset(step, subsetPartition)	
+}
+
+func selectNodesFromOnStep(step Step) []int {
+	if step.Selection != nil {
+		fatal("Two node selection methods on test step")
+	}
+	if step.EndNode == 0 {
+		selected := make([]int, 1)
+		selected[0] = step.OnNode
+		return selected
+	}
+	return makeRange(step.OnNode, step.EndNode)
+}
+
+func selectNodesRange(step Step, config Config) []int{
+	var selection []int
+	if step.Selection.Range.Order == sequential {
+	  if step.Selection.Range.Start <= 0 || 
+			step.Selection.Range.End - step.Selection.Range.Start + 1 > config.Nodes {
+				fatal("Invalid range")
+		}
+		selection = makeRange(step.Selection.Range.Start, step.Selection.Range.End)
+	} else if step.Selection.Range.Order == random {
+		if step.Selection.Range.Number > config.Nodes {
+			fatal("Invalid range")
+		}
+		selection = onePerm(config.Nodes)[0:step.Selection.Range.Number]
+
+	} else /* Invalid order */ {
+			fatal("Invalid Selection format. Order must be SEQUENTIAL or RANDOM ")
+	}
+	return selection
+} 
+
+func selectNodesPercent(step Step, config Config) []int {
+	var percent int
+	var selection []int
+	percent = step.Selection.Percent.Percent
+	if percent > 100 || percent < 0 {
+		fatal("Invalid percent")
+	}
+	numNodes := int( (float64(percent) / 100.0) * float64(config.Nodes) )
+	if step.Selection.Percent.Order == sequential {
+		if step.Selection.Percent.Start -1 + numNodes > config.Nodes {
+			fatal("Invalid start position")
+		}
+		selection = makeRange(step.Selection.Percent.Start, step.Selection.Percent.Start -1 + numNodes)
+	} else if step.Selection.Percent.Order == random {
+		selection = onePerm(config.Nodes)[0:numNodes]
+	} else /* Invalid order */ {
+		fatal("Invalid Selection format. Order must be SEQUENTIAL or RANDOM ")
+	}
+	return selection
+}
+
+func selectNodesSubset(step Step, subsetPartition map[int][]int) []int{
+	if (subsetPartition == nil) {
+		fatal("Subset specified without specifying partion in header")
+	}
+	selected := make([]int, 0)
+	for _, partition := range step.Selection.Subset.Indices {
+		selected = append(selected, subsetPartition[partition]...)
+	}
+	return selected
+}
+
+
+func partition(config Config) map[int][]int{
+	if config.SubsetPartition == nil {
+		return nil
+	} 
+	partitionMap := make(map[int][]int)
+
+	if config.SubsetPartition.Order == sequential {
```

Comment:
Switch statements seem more suitable than if/elif in this block, since you're comparing the same variable repeatedly.